crates/subspace-wasm-tools/src/lib.rs

```diff
@@ -0,0 +1,43 @@
+use std::{env, fs};
+
+/// Creates a `target_file_name` in `OUT_DIR` that will contain constant `bundle_const_name` with
+/// runtime of `runtime_crate_name` stored in it.
+///
+/// Must be called before Substrate's WASM builder.
+pub fn create_runtime_bundle_inclusion_file(
+    target_dir: &str,
+    runtime_crate_name: &str,
+    bundle_const_name: &str,
+    target_file_name: &str,
+) {
+    // Make correct profile accessible in child processes.
+    env::set_var(
+        "WBUILD_PROFILE",
+        env::var("WBUILD_PROFILE")
+            .unwrap_or_else(|_| env::var("PROFILE").expect("Defined by cargo; qed")),
+    );
+
+    // Create a file that will include execution runtime into consensus runtime
+    let profile = env::var("WBUILD_PROFILE").expect("Set above; qed");
+    let execution_wasm_bundle_path = format!(
+        "{target_dir}/{profile}/wbuild/{runtime_crate_name}/{}.compact.wasm",
```

Comment:
Three files will be generated in the build stage: `.wasm`, `.compact.wasm` and `.compact.compressed.wasm`, the final included wasm blob should be:

```suggestion
        "{target_dir}/{profile}/wbuild/{runtime_crate_name}/{}.compact.compressed.wasm",
```

Reply:
Compressed file is not just WASM, it is zstd-compressed version of WASM file. You'll have to decompress it before use. Also since it is included in the primary runtime, it will be compressed with zstd on top anyway, so I'm not 100% confident we want this.

Maybe it will reduce VM memory usage by a bit, what is your concern here exactly?

Reply:
It's more about the robustness. The default behavior in Substrate is targeted to the compressed wasm, and the nested runtime is still a complete and standalone runtime just like the parent runtime. If we use compressed.wasm for the primary runtime, then theoretically we should also use the compressed.wasm for the nested secondary runtime for consistency and to avoid any potential pitfall in the future IMHO.

I have checked the wasm size and found it interesting:

- Use `.compressed.wasm`

```
.rw-r--r--  356k xuliucheng 30 Mar 09:20 subspace_runtime.compact.compressed.wasm
.rw-r--r--  1.1M xuliucheng 30 Mar 09:20 subspace_runtime.compact.wasm
.rwxr-xr-x  1.3M xuliucheng 30 Mar 09:20 subspace_runtime.wasm
```

- Use `.wasm`

```
.rw-r--r--  288k xuliucheng 30 Mar 09:31 subspace_runtime.compact.compressed.wasm
.rw-r--r--  1.5M xuliucheng 30 Mar 09:31 subspace_runtime.compact.wasm
.rwxr-xr-x  1.7M xuliucheng 30 Mar 09:31 subspace_runtime.wasm
```

So it surprisingly seems to be even negative if something is double compressed, okay, given that the on-chain storage is precious, perhaps we should just `.wasm` instead of `.compressed.wasm` and fix any problems whenever unexpected things occur.

Reply:
That was my expectation since most of the code in both runtimes is the same, compression algorithm should be able to effectively de-duplicate a lot of it. By modifying input we increase amount of difference between the two and hurt compression ratio as the result.